kornia/feature/sold2/sold2_detector.py

```diff
@@ -0,0 +1,763 @@
+            [[self.local_patch_radius, self.local_patch_radius]],
+            device=device, dtype=torch.float32)
+        H_patch_points, W_patch_points = torch.where(patch_mask >= 0)
+        patch_points = torch.cat([H_patch_points[..., None],
+                                  W_patch_points[..., None]], dim=-1)
+        # Fetch the circle region
+        patch_center_dist = torch.sqrt(torch.sum(
+            (patch_points - patch_center) ** 2, dim=-1))
+        patch_points = (patch_points[patch_center_dist
+                        <= self.local_patch_radius, :])
+        # Shift [0, 0] to the center
+        patch_points = patch_points - self.local_patch_radius
+
+        # Construct local patch mask
+        patch_points_shifted = (torch.unsqueeze(cand_points_round, dim=2)
+                                + patch_points[None, None, ...])
+        patch_dist = torch.sqrt(torch.sum((torch.unsqueeze(cand_points, dim=2)
+                                          - patch_points_shifted) ** 2,
+                                          dim=-1))
+        patch_dist_mask = patch_dist < dist_thresh[..., None]
+
+        # Get all points => num_points_center x num_patch_points x 2
+        points_H = torch.clamp(patch_points_shifted[:, :, :, 0], min=0,
+                               max=H - 1).to(torch.long)
+        points_W = torch.clamp(patch_points_shifted[:, :, :, 1], min=0,
+                               max=W - 1).to(torch.long)
+        points = torch.cat([points_H[..., None], points_W[..., None]], dim=-1)
+
+        # Sample the feature (N x 64 x 81)
+        sampled_feat = heatmap[points[:, :, :, 0], points[:, :, :, 1]]
+        # Filtering using the valid mask
+        sampled_feat = sampled_feat * patch_dist_mask.to(torch.float32)
+        if len(sampled_feat) == 0:
+            sampled_feat_lmax = torch.empty(0, 64)
+        else:
+            sampled_feat_lmax, _ = torch.max(sampled_feat, dim=-1)
+
+        return sampled_feat_lmax
+
+
+def line_map_to_segments(junctions: np.ndarray, line_map: np.ndarray) -> np.ndarray:
+    """Convert a line map to a Nx2x2 array of segments."""
+    line_map_tmp = line_map.copy()
+
+    output_segments = np.zeros([0, 2, 2])
+    for idx in range(junctions.shape[0]):
+        # if no connectivity, just skip it
+        if line_map_tmp[idx, :].sum() == 0:
+            continue
+
+        # Record the line segment
+        for idx2 in np.where(line_map_tmp[idx, :] == 1)[0]:
+            p1 = junctions[idx, :]  # HW format
+            p2 = junctions[idx2, :]
+            single_seg = np.concatenate([p1[None, ...], p2[None, ...]],
+                                        axis=0)
+            output_segments = np.concatenate(
+                (output_segments, single_seg[None, ...]), axis=0)
+
+            # Update line_map
+            line_map_tmp[idx, idx2] = 0
+            line_map_tmp[idx2, idx] = 0
+
+    return output_segments
+
+
+def super_nms(prob_predictions: np.ndarray, dist_thresh: float,
```

Comment:
implement in pytorch -- also recently added https://kornia.readthedocs.io/en/latest/geometry.bbox.html?highlight=nms#kornia.geometry.bbox.nms

we could benchmark which it's best, or similarly I did with the face detector, make it easy to plug any (e.g torchvision has a nice c++/cuda compiled and quite fast) version that can be easily interfaced
https://github.com/kornia/kornia/blob/be104a5d783788bd5e37338d721e4caf506b0508/kornia/contrib/face_detection.py#L172

Reply:
I did a small benchmark between the kornia version, the torchvision one, and the one I had in my code. The torchvision one is the fastest, both for CPU and GPU, so I am keeping this one.